### PR TITLE
FftSize adjustments

### LINKED ISSUE #6: Defaults need to be more mobile-friendly
There is a lot of lag on my Android while drawing, resulting in a low-poly path. The FFT size should either be smaller or should adapt somehow.

### PR DESCRIPTION
**Pull request by Daynvheur**
FftSize adjustments depending on hardware capabilities (as requested, should fix #6) at first launch, aiming for less than 25ms.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -47,17 +48,21 @@ function loadLocation() { // Inspiration from https://stackoverflow.com/question
...
                         if (v !== null) { // Restriction to valued keys
                             const w = v && decodeURIComponent(v);
                             switch (k) {
                                 case 'pt':
-                                    { // no-case-declaration
+                                    { // no-case-declaration
                                         const [x, y] = w.split(';');
                                         if (x !== null && y !== null)
-                                            addPoint(Number(x), Number(y), false);
+                                            addPoint(Number(x), Number(y), false);
```
Comment (jnm2) [inline src/index.ts]: You're doing something weird to my line endings again :D

File: src/index.ts [hunk 2]
```diff
@@ -92,22 +102,34 @@ function setLocation() {
...
+    complexity = Math.min(complexityNumber.valueAsNumber, fftUnderSize);
 
     circles = complexityCircles.checked;
+
+    const redrawStart = window.performance.now();
     redraw();
+    const redrawStop = window.performance.now();
+
+    if (autoFft && (redrawStop - redrawStart) > 25 && fftSize > 2) {
+        fftSize /= 2;
```
Comment (jnm2) [inline src/index.ts]: Should we prevent it from becoming fractional or does it not matter?

File: src/index.ts [hunk 3]
```diff
@@ -92,22 +102,34 @@ function setLocation() {
...
+        initControls();
+    } else {
+        parameterSlider.value = Math.min(parameter, fftUnderSize).toString();
+        complexityNumber.value = Math.min(complexity, fftUnderSize).toString();
+    }
```
Comment (jnm2) [inline src/index.ts]: Ingenious!

**Review by jnm2 — CHANGES_REQUESTED** [review]
Could you normalize the line endings? `git add --renormalize .` works for me. Everything should be stored in Git with Unix line endings, so I assume what's happening is that you're adding carriage returns. The `.gitattributes` file I added should have fixed this, but maybe you have to reclone to take advantage of that. I recommend setting the Git line ending settings to their defaults.

File: src/index.ts (hunk 1 above)
Comment (Daynvheur) [reply src/index.ts]: Wow, sorry sir!
Guilt goes to copy-pasting between VS and JsFiddle.

File: src/index.ts (hunk 2 above)
Comment (Daynvheur) [reply src/index.ts]: Looks like fftsize handling does not support fractional values for components.length. :D

Thus, fftsize shouldn't go under 2. Fft object prevents from using non-2 powers.

**Review by Daynvheur — COMMENTED** [review]
I'll keep line endings at `\n` on next commit.

File: src/index.ts (hunk 2 above)
Comment (jnm2) [reply src/index.ts]: What if `fftSize` was set through a URL parameter? Then it might not be a power of 2.

Comment (Daynvheur) [reply src/index.ts]: If fftsize is manually set to something that isn't a power of 2, then `new Fft ()` throws and path is not displayed. :)

File: src/index.ts (hunk 1 above)
Comment (jnm2) [reply src/index.ts]: No need to be sorry! No harm done. I was trying to be funny!

File: src/index.ts (hunk 2 above)
Comment (jnm2) [reply src/index.ts]: That's good enough for me.

File: src/index.ts [hunk 4]
```diff
@@ -47,42 +46,48 @@ function loadLocation() { // Inspiration from https://stackoverflow.com/question
...
+
+                                case 'autofft':
+                                    autoFft = Boolean(Number(w));
                                     break;
                             }
-                        }
+                        }
                     }
-                    break;
+                    break;
```
Comment (jnm2) [inline src/index.ts]: I didn't point these out earlier, sorry. Besides this place, there are two more places lower down shown by GitHub where the line endings are being changed and nothing else. Did `git add --renormalize .` not work for you?

File: src/index.ts [hunk 5]
```diff
@@ -223,7 +248,7 @@ function calculateSortedComponentsFromOutput() {
     components.sort((a, b) => b.magnitude - a.magnitude);
 }
 
-function redraw() {
+function redraw(implexity: number = complexity, intrameter: number = parameter) {
```
Comment (jnm2) [inline src/index.ts]: I'm not familiar with what's going on here. 'Implexity' and 'Intrameter'—do they have English meanings or is there another way we could achieve the same goal?

File: src/index.ts [hunk 6]
```diff
@@ -92,22 +97,38 @@ function setLocation() {
...
+    const fftUnderSize = fftSize - 1, minParameter = Math.min(parameter, fftUnderSize), minComplexity = Math.min(complexity, fftUnderSize);
+    circles = complexityCircles.checked;
 
-    complexityNumber.max = (fftSize - 1).toString();
-    complexity = complexityNumber.valueAsNumber;
+    fft = new FFT(fftSize);
+    input = fft.createComplexArray();
+    output = fft.createComplexArray();
+    pathReinitialization();
+    rawPoints.forEach(pt => addPoint(pt.x, pt.y));
```
Comment (jnm2) [inline src/index.ts]: Will this strategy work if you visit the page on your phone with no path initially present?

**Review by jnm2 — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 5 above)
Comment (Daynvheur) [reply src/index.ts]: Yep, the naming is lame. That was a portmanteau between `internal`+complexity/parameter, only as a proof of concept that indeed requires some more attention.

File: src/index.ts (hunk 6 above)
Comment (Daynvheur) [reply src/index.ts]: The previous line being `fft = new FFT(fftSize);`, the fft will adapt at each try cycle, and the path should then be recomputed accordingly. Even on a phone with no path initially present.

File: src/index.ts (hunk 4 above)
Comment (Daynvheur) [reply src/index.ts]: I thought I changed those `\r\n` into `\n`. :o

Comment (Daynvheur) [reply src/index.ts]: It did renomalize, I commited the changes.

File: src/index.ts (hunk 6 above)
Comment (Daynvheur) [reply src/index.ts]: Ow, if `rawPoints`is null, you say?

Comment (Daynvheur) [reply src/index.ts]: 130ab27

File: src/index.ts (hunk 5 above)
Comment (Daynvheur) [reply src/index.ts]: b2a5a5a - underscored global complexity and parameter names

**Comment by Daynvheur** [comment]
I tried to calculate the best fftSize at various points in the draw flow, but anywhere was it satisfying.
Either when the draw is empty the fftsize is always maxed at infinity, either when more than 4 points, the fftsize tends to 4 (the min value).
I'll commit the sources with functions in their original positions, and ComputeRedraw enabled, only for the ref.

**Comment by jnm2** [comment]
This first time I opened http://localhost:1234 in Chrome and Firefox, it settled on 4 as the appropriate FFT size. After refreshing the page, it worked much better and picked an FFT size of 4096.

I'd like to think of a way to make the first-time experience guaranteed not to choose 4 if the device can eventually go higher smoothly. Or maybe just a minimum of 64?

**Comment by Daynvheur** [comment]
Smoothing requires redraw computation, that in itself requires computation.
I can't do anything really relevant on this topic for now, need to have a big rethought on this, before coming back with a fresh idea. :)

**Comment by jnm2** [comment]
That's okay, I don't have any great ideas right now either.

**Review by Daynvheur — COMMENTED** [review]
File review state is outdated

**Comment by Daynvheur** [comment]
Asking to close this PR, as its state is stale. #6 is still under consideration, though the biggest threshold lies behind the complexity path redraw.

**Pull request closed without merging**